### ISSUE TITLE
Add mechanism for listening to LISTEN messages without having to execute anything

### ISSUE THREAD

**Issue by jgdavey** [igrishaev/pg2#28]
Currently, Notification messages only get read from the input buffer after having sent a message to the server first. 

One use-case for listening for messages is to have a dedicated thread and/or connection that subscribes to channel(s) and then dispatches based on the messages that are received. This is a primary reason why the [impossibl pg jdbc](http://impossibl.github.io/pgjdbc-ng/) driver exists: unlike the "official" jdbc driver, this one does not require busy-looping a `SELECT 1` query to actually receive the notify messages.

I played around with a new method on `Connection` called `listenLoop` that would block on just reading from the server, without having to first issue a query:

```java
    public Future<?> listenLoop () {
        flush();
        try (TryLock ignored = lock.get()) {
            return Agent.soloExecutor.submit(() -> {
                    while (!Thread.currentThread().isInterrupted()) {
                        try {
                            final IServerMessage msg = readMessage(false);
                            if (Debug.isON) {
                                Debug.debug(" -> %s", msg);
                            }
                            handleMessage(msg, null);
                        } catch (PGError e) {
                            Debug.debug(" IO issue %s", e);
                            if (isClosed()) { break; }                        
                        } catch (Throwable caught) {
                            break;
                        }
                    }
                    return true;
                });
        }
    }
```

But there are few issues with this:

1. It doesn't allow thread interruption for blocking IO, instead having to wait for socket timeout before the interruption would be seen
2. The details of connection-sharing across threads is murky at best.

What do you think about this?

**Comment by igrishaev** [comment]
Maybe, it could be even simpler with `java.util.TimerTask`. In both cases, the idea is simple: there is Class that accepts a pool/connection, and spawns a regular job that performs an empty select of that connection/pool. There is an API/method to stop the object.

**Comment by igrishaev** [comment]
As far as I understand, between two scheduled calls, a connection or a pool will be free (not used by any thread) and thus can be reused in other threads.

**Comment by igrishaev** [comment]
A very draft:

```java
package org.pg.pubsub;

import org.pg.Connection;
import org.pg.Pool;

import java.util.Timer;
import java.util.TimerTask;

public class Pub implements AutoCloseable {
    private final TimerTask task;
    private final Timer timer;

    private Pub (final TimerTask task, final Timer timer) {
        this.task = task;
        this.timer = timer;
    }
    
    private Pub start () {
        timer.schedule(task, 5000);
        return this;
    }

    public static Pub create(final Pool pool) {
        final Pub pub = new Pub(new TimerTask() {
            @Override
            public void run() {
                final Connection conn = pool.borrowConnection();
                conn.query("select");
                pool.returnConnection(conn);
            }
        }, new Timer());
        return pub.start();
    }

    @Override
    public void close() {
        this.task.cancel();
    }
}
```

What do you think?

**Comment by igrishaev** [comment]
My idea is to have a "hub" object with `.subscribe`, `.unsubscribe`, `.send` methods and similar. One should pass a pool, a minimum config (like what to listen) and then use it for cross-communication.

**Comment by igrishaev** [comment]
It has some difficulties though. For example, since connections are constantly re-created, that object should automatically subscribe them on a given list of topics.

In general, I have an outline idea but will deliver it a bit later. At the moment, I'm working on a prepared statement case (see the `better-ps-cache` branch)

**Comment by jgdavey** [comment]
One other thought here. What about providing a low-level polling method so that user's could decide how/when to set up notification polling. Something that would be lighter weight than the typical `execute` methods:

```java
    public boolean pollNotification() {
        try (TryLock ignored = lock.get()) {
            sendSync();
            sendFlush();
            if (IOTool.available(inStream) < 1) {
                return false;
            }
            final IServerMessage msg = readMessage(false);
            handleMessage(msg, null);
            return true;
        }
    }
```

This would return `true` when a notification was dealt with, false otherwise. 

As an example use-case from a user's perspective, then, you might set up a thread with either a dedicated connection, or one that checks out from a pool, and then runs this continuously until it returns false (in case of many notifications), then sleeps for some amount of time when false.

**Comment by igrishaev** [comment]
That method looks interesting, let me check out. I'll try to simulate this in integration tests and post my updates.

In general, I'm thinkig about building a message queue based on PG2. It should be like Kafka (topics + offets) or SNS/SQL (queues). The problem with listen/notify is, a client won't receive messages until he or she subsribes to a channel. Thus, if someone produced messages 1, 2, 3, and you've subscribed afther the 2, you'll get only 3, and 1 and 2 will be lost. This is a real issue with listen/notify.

**Comment by igrishaev** [comment]
A small update: you can track the changes in this branch: https://github.com/igrishaev/pg2/compare/poll-notifications?expand=1

Brifely, I added a method called `pollUpdates`, but I'm still not sure if it's reliable. The thing is, we should implement a second switch/case across message types, and it's a bit fragile. I thing that after all, it's the same as running an empty query:

```
(pg/query pg "")
```

**Comment by igrishaev** [comment]
Let me check how it behaives with more tests.

**Comment by igrishaev** [comment]
@jgdavey could you please try it using a snapshot? `[0.1.28-SNAPSHOT]`

If briefly, polling messages with `Flush` and `Sync` didn't work because it breaks the flow of messages.

I added a new function `(pg/poll-notifications conn)` that emits an empty query to trigger notification delivery. It returns the number of notifications processed. Postgres handles the case when the query is empty and replies immediately with `NoData` message.

The diff (see the new test): https://github.com/igrishaev/pg2/compare/poll-notifications?expand=1

**Comment by jgdavey** [comment]
This is very interesting, and I'll try this out.

> In general, I'm thinkig about building a message queue based on PG2. It should be like Kafka (topics + offets) or SNS/SQL (queues). The problem with listen/notify is, a client won't receive messages until he or she subsribes to a channel. Thus, if someone produced messages 1, 2, 3, and you've subscribed afther the 2, you'll get only 3, and 1 and 2 will be lost. This is a real issue with listen/notify.

I've done an event system like this in the past using Postgres, and it did use listen/notify, as well as a primary `events` table for the "log". Additionally, it required a `subscriptions` table to keep track of the offsets for named subscriptions. I used LISTEN/NOTIFY just to send "there's new stuff" kind of message to active subscriptions, who would then query to get the new messages

**Comment by igrishaev** [comment]
Just a note: today, I've overwritten the `0.1.28-SNAPSHOT` release with another feature I'm working on in parallel. 

You can try the `(pg/poll-notifications conn)` function by installing pg2 from this branch: `poll-notifications`

Just clone it and run `make install`.

**Comment by igrishaev** [comment]
Regarding a message queue system implemented on top of Postgres, this is what I have in mind.

- it should be free from listen-notify, I think. This is mostly because listeners never receive messages sent BEFORE they subscribed.
- Kafka-like approach: there are producers and consumers. Producers write messages to topics. Consumers read data from topics, and commit their offsets. There is a table that tracks offsets for (topic, consumer) pair. By resetting offsets, messages can be read again.
- SQS/SNS approach: messages are written to queue tables. Reading messages means deleting them (1 query: delete from select).

**Comment by igrishaev** [comment]
I merged the branch `poll-notifications` to master, it's available in `0.1.28`.

To sum up, there is a `(pg/poll-notifications conn)` function that makes an empty query which, in turn, triggers pending notifications to be delivered.

Please check out the test `test-client-test-poll-updates` for more details, I believe it covers all cases.

If you don't mind, I'll close this issue as the basic requirement is met, I think. By calling the new function periodically, you can trigger pending notifications.

I definitely want to build a pubsub system like Kafka or SQS on top of pg2, but let's do it in a separate issue.

**Comment by jgdavey** [comment]
Thanks, I've tried it out, and it definitely feels more correct than sending `select 1` over and over.
